feat(flex-stacker): Add the ability to configure and get histogram measurements from the TOF sensors.

vegano1 opened the pull request:
# Overview

We use the TOF sensor's histogram measurements to determine if there is labware or no labware in a particular location, to do that we need to configure the SPAD maps, as well as some additional settings to make histogram measurements. This pull request adds that configuration and a mechanism to read and stream the Histogram measurements via gcodes.

Closes: [EXEC-1176](https://opentrons.atlassian.net/browse/EXEC-1176) [EXEC-1178](https://opentrons.atlassian.net/browse/EXEC-1178) [EXEC-1181](https://opentrons.atlassian.net/browse/EXEC-1181)

# Changelog

- Add mechanism to configure the TOF sensors SPAD maps + histogram measurements on startup
- Add mechanism to start and stream the TOF Histogram measurements via gcodes
- Change the `i2c_read` API to take an external buffer instead of returning a buffer, this makes streaming large blocks of data (like histogram readings) easier to deal with.


# Review requests

Am I missing something?


[EXEC-1176]: https://opentrons.atlassian.net/browse/EXEC-1176?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[EXEC-1178]: https://opentrons.atlassian.net/browse/EXEC-1178?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[EXEC-1181]: https://opentrons.atlassian.net/browse/EXEC-1181?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[EXEC-1263]: https://opentrons.atlassian.net/browse/EXEC-1263?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

vegano1 committed:
add new GetTOFSensorStatus (M215) gcode

vegano1 committed:
format

vegano1 committed:
add TOFSensorDriverTask scaffolding, GetTOFRegister, SetTOFRegister messages

vegano1 committed:
add TOFSensorTask scaffolding

vegano1 committed:
format

vegano1 committed:
save i2c hw

vegano1 committed:
add i2c_hardare

vegano1 committed:
add TOFDriverPolicy

vegano1 committed:
hook up SetTOFDriver gcode

vegano1 committed:
add transmit_receive i2c method

vegano1 committed:
use I2CComms instead of TOFDriverPolicy in TOFDriverTask

vegano1 committed:
add write protect functions for eeprom and tof sensors

vegano1 committed:
Merge branch 'edge' into tof-from-scratch

vegano1 committed:
Merge branch 'edge' into tof-from-scratch

vegano1 committed:
debug i2c line

vegano1 committed:
hallelujah! i2c comms work!

vegano1 committed:
move i2c setup to main

vegano1 committed:
save

vegano1 committed:
fixed the i2c line!
- switched to HAL_I2C_Mem_Read_IT and HAL_I2C_Mem_Write_IT, as i was having issues with Master Rear/Write IT
- added implementation for SetTOFRegister and GetTOFRegister, still need validation

vegano1 committed:
added EnableTOFSensor and implementation

vegano1 committed:
format

vegano1 committed:
rename tof_i2c_hardware and i2c files

vegano1 committed:
save

vegano1 committed:
start adding type to enum convertion for registers

vegano1 committed:
save

vegano1 committed:
Merge branch 'edge' into flex-stacker-tof-sensors-bringup

vegano1 committed:
fix bad merge

vegano1 committed:
tof

vegano1 committed:
remove annoying ^M carriedge return from from files.

vegano1 committed:
add startup tmf8821

vegano1 committed:
save

vegano1 committed:
Merge branch 'edge' into flex-stacker-tof-sensors-bringup

vegano1 committed:
feat(flex-stacker): enable tof sensors

vegano1 committed:
- moved eepriom write pin hardware functions to system.c
- rename tof_write_protect to tof_enable_pin
- change BUFFER to be uint8_t instead of uint32_t.
- cleaning up

vegano1 committed:
format

vegano1 committed:
- added scripts/hex_to_c.py script that generates c array files from hex files.
- added scripts/requirements.txt to keep track of script python dependencies
- added a cmake mechanism to install the scripts/requirements.txt file
- added a cmake mechanism for the flex-stacker that uses the scripts/hex_to_c.py to generate the tmf8820_image.c file from the tmf8820_image.hex file.

vegano1 committed:
format

vegano1 committed:
feat(flex-stacker): add SPAD map configuration on starup.

vegano1 committed:
add common config

vegano1 committed:
add GetTOFHistogram (225) gcode and implementation.

vegano1 committed:
implement get_histogram_measurements

vegano1 committed:
Merge branch 'edge' into flex-stacker-tof-sensor-config

sfoster1 commented on stm32-modules/include/flex-stacker/flex-stacker/tmf8820.hpp:
```diff
@@ -223,6 +284,234 @@ class TMF8820 {
+            auto spad_mask = sensor_id == TOF_X ? spad_mask_x : spad_mask_z;
+            auto spad_mask_len =
+                sensor_id == TOF_X ? spad_mask_x_length : spad_mask_z_length;
+            for (uint8_t i = 0; i < spad_mask_len; i++) {
```
maybe use `memcpy` or `std::copy` or something

sfoster1 commented on stm32-modules/include/flex-stacker/flex-stacker/tmf8820.hpp:
```diff
@@ -223,6 +284,234 @@ class TMF8820 {
+            auto spad_map = sensor_id == TOF_X ? spad_map_x : spad_map_z;
+            auto spad_map_len =
+                sensor_id == TOF_X ? spad_map_x_length : spad_map_z_length;
+            for (uint8_t i = 0; i < spad_map_len; i++) {
```
ditto for the copy

vegano1 committed:
add delay before initializing tof sensors

vegano1 committed:
feat(flex-stacker): add mechanism to read out the histogram over gcode.

sfoster1 commented on stm32-modules/include/flex-stacker/flex-stacker/tof_utils.hpp:
```diff
@@ -0,0 +1,47 @@
+
+inline auto base64_encode(const std::array<uint8_t, HISTOGRAM_LEN>& data)
+    -> std::string {
+    std::string encoded;
```
this kind of use of a `std::string` will heap-allocate its internal storage, possibly multiple times depending on the implementation since we're repeatedly calling `push_back` on it. We should use a cstring or something with a fixed initial size and store the size somewhere or null-terminate it so we don't allocate.

sfoster1 reviewed (changes requested):
It looks good overall but we shouldn't use a `std::string` for this since it allocates internal memory, and creating it with 0 length and then repeatedly calling `push_back` will force a couple reallocations on top of that. We should use a fixed-size buffer.

sfoster1 reviewed (changes requested):
Oh also please add tests, at least for the message sequencing and retry mechanisms. Please also add tests for the base64 code; ideally these would be roundtripping with python, but you could also generate some strings with python, put them in the test file, and check against those.

ahiuchingau commented on stm32-modules/include/flex-stacker/flex-stacker/tof_sensor_task.hpp:
```diff
@@ -184,6 +188,114 @@ class TOFSensorTask {
+            // TODO: send specific error code
+            response = messages::ErrorMessage{
+                .code = errors::ErrorCode::TMC2160_INVALID_ADDRESS};
+            return send_response(response);
```
Do we need to do something here to signal the host comms task to remove the `get_tof_measurement_cache`, so we can receive and process another measurement request?

vegano1 replied:
yeah, good call out.

ahiuchingau reviewed (commented):
Got a question about host comms task
🛫 🛫 🛫

vegano1 committed:
address feedback

vegano1 committed:
working again

vegano1 committed:
parsing works

vegano1 committed:
- change base64_encode to take array for output instead of using std::string, which allocates heap memory.
- propagate the encoded_array from the tof_sensor_task, all the way down tot the base64_encode function so we arent allocating a ton of stack space.

vegano1 committed:
add tof_utils.hpp unit test

vegano1 committed:
one CATCH_CONFIG_MAIN

vegano1 committed:
fix padding issue with base64_encoder

vegano1 committed:
format

vegano1 committed:
Merge branch 'edge' into flex-stacker-tof-sensor-config

vegano1 committed:
Merge branch 'edge' into flex-stacker-tof-sensor-config

sfoster1 commented on stm32-modules/include/flex-stacker/flex-stacker/tof_sensor_task.hpp:
```diff
@@ -173,17 +178,122 @@ class TOFSensorTask {
+    auto reset_measurement_state(TOFSensorID sensor_id) -> void {
+        auto sensor = &get_sensor(sensor_id);
+        sensor->driver.stop_measurement(sensor_id);
+        // sensor->current_frame = std::nullopt;
```
commented out code?

sfoster1 reviewed (approved):
Looks awesome now! One little commented out line of code is all.

sfoster1 commented on stm32-modules/include/flex-stacker/flex-stacker/messages.hpp:
```diff
@@ -3,6 +3,7 @@
 #include <concepts>
 #include <cstdint>
 #include <optional>
+#include <string>
```
I think we don't need this anymore

sfoster1 commented on stm32-modules/include/flex-stacker/flex-stacker/tof_sensor_task.hpp:
```diff
@@ -173,17 +178,122 @@ class TOFSensorTask {
     }
 
+    auto reset_measurement_state(TOFSensorID sensor_id) -> void {
+        auto sensor = &get_sensor(sensor_id);
```
Why take the address and pointerfy this? it returns a reference, so this doesn't do much

vegano1 replied:
cause Im a dummy, thanks for pointing it out.

sfoster1 commented on stm32-modules/include/flex-stacker/flex-stacker/tof_sensor_task.hpp:
```diff
@@ -173,17 +178,122 @@ class TOFSensorTask {
+
+    auto visit_message(const messages::GetTOFMeasurementMessage& m) -> void {
+        messages::HostCommsMessage response;
+        auto sensor = &get_sensor(m.sensor_id);
```
why pointerfy?

sfoster1 commented on stm32-modules/include/flex-stacker/flex-stacker/tof_sensor_task.hpp:
```diff
@@ -173,17 +178,122 @@ class TOFSensorTask {
+
+    auto visit_message(const messages::ManageTOFMeasurementMessage& m) -> void {
+        messages::HostCommsMessage response;
+        auto sensor = &get_sensor(m.sensor_id);
```
why pointer?

ahiuchingau commented on stm32-modules/flex-stacker/tests/test_tof_utils.cpp:
```diff
@@ -0,0 +1,96 @@
+            0x32, 0x33, 0x34, 0x35, 0x36, 0x37, 0x38, 0x39, 0x3A, 0x3B,
+            0x3C, 0x3D, 0x3E, 0x3F, 0x40, 0x41, 0x42, 0x43, 0x44, 0x45};
+        std::string expected =
+            "AAECAwQFBgcICQoLDA0ODxAREhMUFRYXGBkaGxwdHh8gISIjJCUmJygpKissLS4vMD"
```
so cool

vegano1 committed:
address feedback
